bdb/rep.c

```diff
@@ -3729,6 +3732,85 @@ void bdb_set_seqnum(void *in_bdb_state)
 
 int gbl_online_recovery = 1;
 
+static pthread_mutex_t rep_mon_lk = PTHREAD_MUTEX_INITIALIZER;
+static pthread_cond_t rep_mon_cd = PTHREAD_COND_INITIALIZER;
+static pthread_t rep_mon_td;
+
+struct rep_mon {
+    pid_t tid;
+    char *host;
+    int starttime;
+    int type;
+};
+static hash_t *rep_mon_hash = NULL;
+
+int gbl_rep_mon_threshold = 0;
+
+static int process_rep_mon_hash(void *obj, void *arg)
+{
+    struct rep_mon *rm = (struct rep_mon *)obj;
+    int now = comdb2_time_epoch(), threshold = gbl_rep_mon_threshold; 
+    if (threshold > 0 && now - rm->starttime > threshold) {
+        fprintf(stderr, "Thread %u host %s type %d hung for %d seconds\n",
+            rm->tid, rm->host, rm->type, now - rm->starttime);
+        char pstack_cmd[128];
+        snprintf(pstack_cmd, sizeof(pstack_cmd), "pstack %d", (int)rm->tid);
+        system(pstack_cmd);
+        /* Linux pstacks tid, non-linux halts hash-for */
+#ifndef _LINUX_SOURCE
+        return 1;
+#endif
+    }
+    return 0;
+}
+
+static void *rep_mon(void *thd)
+{
+    pthread_mutex_lock(&rep_mon_lk);
+    if (rep_mon_hash) {
+        return NULL;
+    }
+    rep_mon_td = pthread_self();
+    rep_mon_hash = hash_init(sizeof(pthread_t));
+    logmsg(LOGMSG_INFO, "Starting replication monitor thread\n");
+    while(gbl_rep_mon_threshold > 0)
+    {
+        struct timespec waittime;
+        clock_gettime(CLOCK_REALTIME, &waittime);
+        waittime.tv_sec += 1;
+        pthread_cond_timedwait(&rep_mon_cd, &rep_mon_lk, &waittime);
+        if (gbl_rep_mon_threshold > 0) {
+            hash_for(rep_mon_hash, process_rep_mon_hash, NULL);
+        }
+    }
+    hash_free(rep_mon_hash);
+    rep_mon_hash = NULL;
+    logmsg(LOGMSG_INFO, "Exiting replication monitor thread\n");
+    pthread_mutex_unlock(&rep_mon_lk);
+    return NULL;
+}
+
+static void add_rep_mon(struct rep_mon *rm)
+{
```

Comment:
Wondering if we should NOOP this and `rem_rep_mon` functions for the default case `gbl_rep_mon_threshold == 0`, and then we could change the below `else if` into just an `else`.

Reply:
I think we have to use only the hash actually- imagine that the tunable is disabled after something is added to the hash .. because it's kept on the stack, the rep_mon thread might be looking at free'd stack memory if rem_rep_mon doesn't remove it ..

Reply:
Maybe we can use gbl_rep_mon_threshold for add_rep_mon ..

Reply:
Okay .. I've introduced an 'added' field to struct rep_mon .. if it is added to the rep_mon hash, this will be set to 1.  rem_rep_mon gets the mutex only if added is set to 1.